src/sqlfluff/rules/L065.py

```diff
@@ -0,0 +1,148 @@
+    document_fix_compatible,
+    document_groups,
+)
+
+
+@document_groups
+@document_fix_compatible
+class Rule_L065(BaseRule):
+    """Set operators should be surrounded by newlines.
+
+    **Anti-pattern**
+
+    In this example, `UNION ALL` is not on a line ifself.
+
+    .. code-block:: sql
+
+        SELECT 'a' AS col UNION ALL
+        SELECT 'b' AS col
+
+    **Best practice**
+
+    .. code-block:: sql
+
+        SELECT 'a' AS col
+        UNION ALL
+        SELECT 'b' AS col
+
+    """
+
+    groups = ("all",)
+
+    _target_elems = ("set_operator",)
+
+    def _eval(self, context: RuleContext) -> List[LintResult]:
+        """Set operators should be surrounded by newlines.
+
+        For any set operator we check if there is any NewLineSegment in the non-code
+        segments preceeding or following it.
+
+        In particular, as part of this rule we allow multiple NewLineSegments.
+        """
+        segment = context.functional.segment
+
+        expression = segment.children()
+        set_operator_segments = segment.children(sp.is_type(*self._target_elems))
+        results: List[LintResult] = []
+
+        # If len(set_operator) == 0 this will essentially not run
+        for set_operator in set_operator_segments:
+            preceeding_code = (
+                expression.reversed().select(start_seg=set_operator).first(sp.is_code())
+            )
+            following_code = expression.select(start_seg=set_operator).first(
+                sp.is_code()
+            )
+            res = {
+                "before": expression.select(
+                    start_seg=preceeding_code.get(), stop_seg=set_operator
+                ),
+                "after": expression.select(
+                    start_seg=set_operator, stop_seg=following_code.get()
+                ),
+            }
+
+            newline_before_set_operator = res["before"].first(sp.is_type("newline"))
+            newline_after_set_operator = res["after"].first(sp.is_type("newline"))
+
+            # In the case of when there is no newline at all in the preceeding/following
+            # segments this is the directly preceeding/following whitespace character
+            preceeding_whitespace = res["before"].first(sp.is_whitespace()).get()
+            following_whitespace = res["after"].first(sp.is_whitespace()).get()
+
+            if newline_before_set_operator and newline_after_set_operator:
+                continue
+            elif not newline_before_set_operator and newline_after_set_operator:
+                results.append(
+                    LintResult(
+                        anchor=set_operator,
+                        description=(
+                            "Set operators should be surrounded by newlines. "
+                            f"Missing newline before set operator {set_operator.raw}."
+                        ),
+                        fixes=[
+                            LintFix.replace(
```

Comment:
Critical CI Test is complaining about 

```
self = <[AttributeError("'LintFix' object has no attribute 'edit'") raised in repr()] LintFix object at 0x106bc5330>, edit_type = 'replace', anchor = None, edit = [<NewlineSegment: (None) '\n'>], source = None

    def __init__(
        self,
        edit_type: str,
        anchor: BaseSegment,
        edit: Optional[Iterable[BaseSegment]] = None,
        source: Optional[Iterable[BaseSegment]] = None,
    ) -> None:
        if edit_type not in (
            "create_before",
            "create_after",
            "replace",
            "delete",
        ):  # pragma: no cover
            raise ValueError(f"Unexpected edit_type: {edit_type}")
        self.edit_type = edit_type
        if not anchor:  # pragma: no cover
>           raise ValueError("Fixes must provide an anchor.")
E           ValueError: Fixes must provide an anchor.

.tox/ruleslinting/lib/python3.10/site-packages/sqlfluff/core/rules/base.py:154: ValueError
```

here.

I'm not really sure what the actual issue is, I copied this from other rules where it seems to work fine 🤔

Reply:
I think the issue here is, that for `tsql` the preceeding newline is part of the `from_expressions`.

```
[L:  8, P:  3]      |                    from_clause:
[L:  8, P:  3]      |                        keyword:                              'FROM'
[L:  8, P:  7]      |                        whitespace:                           ' '
[L:  8, P:  8]      |                        from_expression:
[L:  8, P:  8]      |                            [META] indent:
[L:  8, P:  8]      |                            from_expression_element:
[L:  8, P:  8]      |                                table_expression:
[L:  8, P:  8]      |                                    table_reference:
[L:  8, P:  8]      |                                        identifier:           'dbo'
[L:  8, P: 11]      |                                        dot:                  '.'
[L:  8, P: 12]      |                                        identifier:           'SUPPLY3'
[L:  8, P: 19]      |                            newline:                          '\n'
[L:  9, P:  1]      |                            [META] dedent:
[L:  9, P:  1]      |                set_operator:
[L:  9, P:  1]      |                    keyword:                                  'UNION'
[L:  9, P:  6]      |                    whitespace:                               ' '
[L:  9, P:  7]      |                    keyword:                                  'ALL'
[L:  9, P: 10]      |                newline:                                      '\n'
```

In this rule I assume that set operators are immediately proceeded by either Newline of Whitespace. 

Would it be possible to change this behaviour in `tsql` ?

Reply:
@barrywhart could you help out here?

Reply:
Do you still need my input on this?

Newlines are often at different levels of the hierarchy; in general you can't count on them to be siblings of a particular segment. You can often use `raw_segment_pre:` and/or `raw_stack` to find the previous raw segment.